ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureBlobDataAccessor.java

```diff
@@ -151,123 +152,100 @@ ProxyOptions getProxyOptions() {
    * @param cloudBlobMetadata the blob metadata
    * @param blobInputStream the input stream
-   * @return {@code true} if the upload was successful, {@code false} if the blob already exists.
-   * @throws BlobStorageException for any error on ABS side.
-   * @throws IOException for any error with supplied data stream.
+   * @return a CompletableFuture of type {@link Boolean} that will eventually be {@code true} if the upload was successful,
+   *         or {@code false} if the blob already exists. If there was an error on ABS side, it will complete exceptionally
+   *         containing the BlobStorageException.
    */
-  public boolean uploadIfNotExists(BlobId blobId, long inputLength, CloudBlobMetadata cloudBlobMetadata,
-      InputStream blobInputStream) throws BlobStorageException, IOException {
+  public CompletableFuture<Boolean> uploadAsyncIfNotExists(BlobId blobId, long inputLength,
+      CloudBlobMetadata cloudBlobMetadata, InputStream blobInputStream) {
     BlobRequestConditions blobRequestConditions = new BlobRequestConditions().setIfNoneMatch("*");
     azureMetrics.blobUploadRequestCount.inc();
     Timer.Context storageTimer = azureMetrics.blobUploadTime.time();
-    try {
-      Map<String, String> metadata = cloudBlobMetadata.toMap();
-      storageClient.uploadWithResponse(blobId, blobInputStream, inputLength, null, metadata, null, null,
-          blobRequestConditions).get(uploadTimeout.toMillis(), TimeUnit.MILLISECONDS);
-      logger.debug("Uploaded blob {} to ABS", blobId);
-      azureMetrics.blobUploadSuccessCount.inc();
-      return true;
-    } catch (ExecutionException | InterruptedException | TimeoutException e) {
-      Exception ex = Utils.extractFutureExceptionCause(e);
-      if (ex instanceof BlobStorageException) {
-        BlobStorageException bse = (BlobStorageException) ex;
-        if (bse.getErrorCode() == BlobErrorCode.BLOB_ALREADY_EXISTS) {
+    final CompletableFuture<Void> responseFuture = FutureUtils.orTimeout(
+        storageClient.uploadWithResponse(blobId, blobInputStream, inputLength, null, cloudBlobMetadata.toMap(), null,
+            null, blobRequestConditions), uploadTimeout);
+    return responseFuture.handle((response, throwable) -> {
+      storageTimer.stop();
+      if (throwable != null) {
+        Exception ex = Utils.extractFutureExceptionCause(throwable);
+        if (ex instanceof BlobStorageException
+            && ((BlobStorageException) ex).getErrorCode() == BlobErrorCode.BLOB_ALREADY_EXISTS) {
+          // Received exception due to blob being already present
           logger.debug("Skipped upload of existing blob {}", blobId);
           azureMetrics.blobUploadConflictCount.inc();
           return false;
-        } else {
-          throw bse;
         }
-      } else if (ex instanceof UncheckedIOException) {
-        // error processing input stream
-        throw ((UncheckedIOException) ex).getCause();
+        throw new CompletionException(ex);
```

Comment:
It looks like we are only able to re-throw unchecked exceptions from the completion stages. Since exceptions from previous stages are of `Throwable` type which can be checked exceptions as well, it doesn't allow me to re-throw unless we wrap it inside `CompletionException`.